File: trees/sum_root_to_leaf_numbers.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Integer}
def sum_numbers(root)
    return 0 if root.nil?

    total_sum = []
    
    def collect_paths(node, number, total_sum)
        return if node.nil?

        number = number + node.val.to_s

        if node.left.nil? && node.right.nil?
            total_sum << number.to_i
            return
        end

        collect_paths(node.left, number, total_sum) if node.left
        collect_paths(node.right, number, total_sum) if node.right

    end

    collect_paths(root, "", total_sum)

    return total_sum.reduce(:+)

end

File: trees/sum_root_to_leaf_numbers.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sumNumbers(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0

        total_sum = []

        def collect_paths(node: TreeNode, number: str, total_sum: List[int]) -> None:
            if not node:
                return

            number = number + str(node.val)

            if not node.left and not node.right:
                total_sum.append(int(number))
                return

            if node.left:
                collect_paths(node.left, number, total_sum)
            if node.right:
                collect_paths(node.right, number, total_sum)

        collect_paths(root, "", total_sum)

        return sum(total_sum)